keps/sig-api-machinery/4153-declarative-validation/README.md

```diff
@@ -0,0 +1,1503 @@
+of the versioned types to be the hub type? The vast majority of APIs only have
+one version anyway. The obvious candidate version to choose for the hub version
+would the perferred storage version.
+
+Switching to a versioned type for the hub type would have a few implications:
+
+- We would eliminate the need for internal versions.
+- We would reduce the need for "unsafe" conversions (a small security win) since
+  most APIs only have a single v1 version and so the entire request handling process
+  would simply use that version.
+- We would introduce more conversion when API request version differs from the
+  hub version. But beta APIs are off-by-default and we expect a lot less mixed
+  version API usage than in the past.
+
+This "hub version" change feels like something that could be made somewhat
+independant of this KEP.
+
+
+##### Mitigation: Avoid Conversion to Internal Type
+
+Requests are received as the versioned type, so it should be feasible to avoid
+extra conversions for resources that have no need of handwritten validations.
+
+##### Mitigation: Avoid Unstructured Conversion
+
+SMD's [value](https://github.com/kubernetes-sigs/structured-merge-diff/tree/master/value)
+wrappers have already been benchmarked and shown to have low overhead.
+
+##### Mitigation: Use Value Validations Where Possible
+
+Most validations will be possible to evaluate with simple value validations
+in native OpenAPI (required checks, enum value checks, etc). The rest are
+CEL expressions which are expected to account for 30% of all validation rule.
+
+CEL evaluation is relatively efficient we will gather detailed benchmarks baselined against native types to demonstrate this.
+
+##### Mitigation: Go fast paths for hot validation code paths
+
+If CEL becomes a bottleneck, we may explore possibilities to run Go rather than
+interpreted CEL behind the scenes for certain hot code paths:
+  - Generate Go code from CEL expressions.
+  - Hand write validation code for select validations and call those
+    directly (or from CEL).
+
+## Design Details
+
+### Phase 1: Expand Schema Capability
+1. Introduce Feature Flag: `OpenAPIV3CELValidations` which would expose
+`x-kubernetes-validations` in OpenAPIV3 spec for native types.
+2. Introduce new [CEL libraries](#cel-validation) that will be needed to verify native schemas.
+3. Add declarative defaulting support to OpenAPI code generator
+4. Create a test framework to validate objects against internal and versioned types
+using the handwritten and schema-based validation backends, respectively.
+5. Create library for validating native resources against their OpenAPI Schemas.
+
+### Phase 2: Enrichment of schemas
+ 
+Publish a tracking issue with example of a group-version being annotated, and a guide for contributors to:
+1. Add declarative validation rules using [Go IDL tags](#proposal).
+1. Instrument validation_test.go unit tests to ensure declarative
```

Comment:
Why not always run the declarative validation and have a log when it has an error but regular validation doesn't? And probably vice-versa? As a way to detect some differences between the two.

Reply:
I'm not sure I follow. How is this different from the current proposal to instrument the tests to run against all forms of validation and report errors

Reply:
This is live, for the cases that maybe aren't covered by the tests?

Reply:
Yes that is written in the design details that during alpha we keep returning the old sets of errors but log any differences while the feature is enabled

Reply:
Thanks, sorry I missed it 👍